### PR TITLE
fix bug 1148498; wrapper and split roles

### PR DESCRIPTION
**Pull request by phrawzty**
ping @rhelmer

### CODE REVIEW

File: terraform/wrapper.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+fi
+
+# The secret_bucket is specified here.
+if [ ! -r $TFVARS ]; then
+    echo "ERROR: Could not read ${TFVARS}"
+    exit 1
+fi
+
+BUCKET=$(grep secret_bucket terraform.tfvars | awk -F\" '{print $2}')
+ENV=$1
```
Comment (rhelmer) [inline terraform/wrapper.sh]: maybe we should take another arg here like:
```
INFRA=$3
```
That could be ```.``` by default, or you could pass an arg like ```./wrapper.sh env ./consul```

File: terraform/wrapper.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
```
Comment (rhelmer) [inline terraform/wrapper.sh]: If we go with the ```$INFRA``` idea above, could do a ```pushd $INFRA``` here

File: terraform/wrapper.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# This script should be run as a wrapper to Terraform.
+# It should be symlinked in consul/ as well.
+
+TFVARS="terraform.tfvars"
+
+# Bail on errors and be verbose.
+set -ex
+
+# Need to specify the environment.
+if [ "${1}x" == "x" ]; then
```
Comment (rhelmer) [inline terraform/wrapper.sh]: maybe ```[ $# != 2 ]``` (or whatever expected number of args are)

File: terraform/wrapper.sh (hunk 1 above)
Comment (rhelmer) [inline terraform/wrapper.sh]: Also we don't need to do this now, but it'd behoove us to break the ```main.tf``` up further into subdirectories, as we have done with ```./consul``` - ```collector```, ```elasticsearch```, ```webapp``` and so on, and not have a ```./terraform/main.tf``` in here at all.

**Comment by phrawzty** [comment]
This requires that we split out the roles into distinct TF units (as we've discussed previously), so while this script is good, it doesn't *currently* do anything useful. :smile:

**Comment by phrawzty** [comment]
Ok, this splits out all of the roles, which comes with the ugly side effect of creating a tonne of (nearly) duplicate resources all over the goddamn place. :ghost: The wrapper is good for now, but I suspect that we might end up replacing it with a Makefile or something at some point, heh.

@rhelmer `r?`

**Comment by rhelmer** [comment]
The ```README.md``` should mention these changes

**Comment by rhelmer** [comment]
Should there be a symlink in each dir too for ```terraform.tfvars``` ?

**Comment by phrawzty** [comment]
> The README.md should mention these changes

Agreed.

> Should there be a symlink in each dir too for terraform.tfvars ?

`terraform.tfvars` is in `.gitignore`, so that's not possible.  We could add that to the wrapper (now you see why I'm already envisioning a Makefile, hehe).

**Comment by rhelmer** [comment]
> terraform.tfvars is in .gitignore, so that's not possible. We could add that to the wrapper (now you see why I'm already envisioning a Makefile, hehe).

Ah, yeah I think it's starting to make a lot of sense.

File: terraform/wrapper.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+# Grab tfstate from S3
+aws s3 sync --exclude="*" --include="terraform.tfstate" s3://${BUCKET}/tfstate/${ENV}/ ./
+
+# Run TF
+pushd $ROLE
+terraform apply -var "environment=${ENV}"
+popd
+
+# Upload tfstate to S3
+aws s3 sync --exclude="*" --include="terraform.tfstate" ./ s3://${BUCKET}/tfstate/${ENV}/
```
Comment (rhelmer) [inline terraform/wrapper.sh]: Do we care about updating this after ```terraform destroy```?

Comment (phrawzty) [inline terraform/wrapper.sh]: That's a good point; yes, we probably do.  So I guess we're now talking about:
```bash
    echo "USAGE: ${0} <action> <environment> <role>"
```

File: terraform/wrapper.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+# Pre-flight check is good, let's continue.
+ENV=$1
+ROLE=$2
+
+# Be verbose and bail on errors.
+set -ex
+
+# Grab tfstate from S3
+aws s3 sync --exclude="*" --include="terraform.tfstate" s3://${BUCKET}/tfstate/${ENV}/ ./
```
Comment (rhelmer) [inline terraform/wrapper.sh]: doesn't this need to happen inside the ```pushd``` below? Can there be one ```terraform.tfstate``` for everything?

Comment (phrawzty) [inline terraform/wrapper.sh]: Oops, yes, sorry, I fixed that but forgot to commit. :blush:

File: terraform/wrapper.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+# Pre-flight check is good, let's continue.
+ENV=$1
+ROLE=$2
+
+# Be verbose and bail on errors.
+set -ex
+
+# Grab tfstate from S3
+pushd $ROLE
+aws s3 sync --exclude="*" --include="terraform.tfstate" s3://${BUCKET}/tfstate/${ENV}/ ./
```
Comment (rhelmer) [inline terraform/wrapper.sh]: Do we need to have ```${ROLE}``` in the S3 path too?

File: terraform/wrapper.sh [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+# Run TF
+terraform apply -var "environment=${ENV}"
+
+# Upload tfstate to S3
+aws s3 sync --exclude="*" --include="terraform.tfstate" ./ s3://${BUCKET}/tfstate/${ENV}/
```
Comment (rhelmer) [inline terraform/wrapper.sh]: Ditto above

**Comment by rhelmer** [comment]
r+

**Pull request merged**